core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -701,14 +708,22 @@ private[kafka] class Processor(val id: Int,
         openOrClosingChannel(receive.source) match {
           case Some(channel) =>
             val header = RequestHeader.parse(receive.payload)
-            val connectionId = receive.source
-            val context = new RequestContext(header, connectionId, channel.socketAddress,
-              channel.principal, listenerName, securityProtocol)
-            val req = new RequestChannel.Request(processor = id, context = context,
-              startTimeNanos = time.nanoseconds, memoryPool, receive.payload, requestChannel.metrics)
-            requestChannel.sendRequest(req)
-            selector.mute(connectionId)
-            handleChannelMuteEvent(connectionId, ChannelMuteEvent.REQUEST_RECEIVED)
+            if (header.apiKey() == ApiKeys.SASL_HANDSHAKE && channel.maybeBeginServerReauthentication(receive, time.nanoseconds))
+              trace(s"Begin re-authentication: $channel")
+            else if (channel.serverAuthenticationSessionExpired(time.nanoseconds)) {
```

Comment:
Ok, it's now just a single call to `time.nanoseconds()` under all circumstances except for when `SASL_HANDSHAKE_REQUEST` is sent and either 1) re-authentication is not enabled; or 2) re-authentication is enabled but it occurred less than a second ago.  In these two cases `time.nanoseconds()` will be invoked twice, but it probably doesn't matter since this is a rare occurrence and we are going to send an error back to the client if/when it happens.